[Alex] step-2 체스판 생성

horanga opened the pull request:
## 구현 내용
-추후에 board에 Pawn 말고 다른 Piece들도 들어올 것을 대비해서 Pawn 클래스를 상위클래스인 Piece를 상속하도록 했고, Board의 메서드들도 인자를 Piece로 받는 방식을 구현함

-Pawn 인스턴스들이 동일한지 확인하기 위해서 hascode와 equals 자동 오버라이드 기능을 활용했음

## 고민 사항
테스트에서 중복된 부분을 before 애노테이션을 사용하라는 요구사항이 있었는데, 현재 상황에서는 board라는 컬렉션에 Pawn인스턴스를 넣는 부분이 중복됐음. 그런데, 이걸 beforeEach를 쓰려면 board라는 컬렉션이 전역적으로 접근할 수 있는 필드변수가 돼야 한다. 이렇게 되면 테스트간에 서로 간섭을 하게 되기 때문에 테스트 구조 측면에서 좋지 않다고 판단을 했음.

-Pawn 말고 다른 데이터 타입이 들어오면 컴파일 에러가 뜨게끔 구성하라고 해서 인자를 상위클래스 타입인 Pices로 설정해둠. 근데 컴파일 에러가 뜨는 걸 테스트하는 방법은 없는 거 같아서 코드는 각주 처리로 작성해두기만 함

## 기타

horanga committed:
Revert "feat: 이눔 컬러 생성"

This reverts commit d5852e1d816a55aa3c30977128dd30da8a4ddacd.

horanga committed:
feat:Pawn 레코드 생성
불변객체를 만들기 위해서 레코드를 생성

horanga committed:
refactor: Pawn 레코드에서 클래스로 수정
불변객체이면 안돼서 클래스로 수정

feat: Board 클래스 생성
Pieces 객체들을 넣는 보드 생성

refator: Pawn을 record에서 class로 변경
불변객체일 필요가 없어서 class로 변경

feat:보드에 Pieces 객체 넣는 로직 구현
pieces들을 담아놓는 board 필요. 이 안에서 여러 로직을 구현할 계획

refactor: Pawn 테스트 변경
Pawn 클래스 변경에 따른 테스트 수정

Revert "feat: 이눔 컬러 생성"

This reverts commit d5852e1d816a55aa3c30977128dd30da8a4ddacd.

feat:Pawn 레코드 생성
불변객체를 만들기 위해서 레코드를 생성

feat: push 취소

horanga committed:
refactor: Pawn 레코드에서 클래스로 수정
불변객체이면 안돼서 클래스로 수정

horanga committed:
feat: push 취소

horanga committed:
feat: Board 클래스 생성
Pieces 객체들을 넣는 보드 생성

horanga committed:
refator: Pawn을 record에서 class로 변경
불변객체일 필요가 없어서 class로 변경

horanga committed:
feat:보드에 Pieces 객체 넣는 로직 구현
pieces들을 담아놓는 board 필요. 이 안에서 여러 로직을 구현할 계획

horanga committed:
refactor: Pawn 테스트 변경
Pawn 클래스 변경에 따른 테스트 수정

horanga committed:
Merge branch 'step-2' of https://github.com/horanga/be-chess into step-2

horanga committed:
Merge branch 'step-2' of https://github.com/horanga/be-chess into step-2

honux77 commented on untitled/src/main/java/org/example/chess/Board.java:
```diff
@@ -0,0 +1,24 @@
+package org.example.chess;
```
패키지 이름에서 org.example은 제거해 주세요.

honux77 commented on untitled/src/main/java/org/example/chess/Board.java:
```diff
@@ -0,0 +1,24 @@
+        board.add(piece);
+    }
+
+    public Piece findPiece(int number){
```
number 보다는 index나 location 같은 이름이 좋을 것 같습니다.

honux77 reviewed (changes requested):
지금 최상위 폴더가 untitled 도 되어 있는데 하나씩 상위로 옮겨 주세요. 
그리고 기타 코멘트도 확인해 주세요.

honux77 commented on untitled/src/main/java/org/example/piece/Pawn.java:
```diff
@@ -0,0 +1,35 @@
+    public static final String BLACK = "black";
+    public static final String WHTIE = "white";
+
+    String color;
```
특별한 이유가 없다면 private을 접근 제한자로 사용해 주세요.

honux77 commented on untitled/src/main/java/org/example/piece/Pawn.java:
```diff
@@ -0,0 +1,35 @@
+    String color;
+
+    public Pawn(String color){
+        if(!color.equals(BLACK)&&!color.equals(WHTIE)){
```
코드 포매팅도 신경써 주세요.

horanga committed:
refactor: 변경사항 반영

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static List<List<Piece>> intialize() {
+        List<List<Piece>> board = new ArrayList<>();
+        for (int i = 0; i < 8; i++) {
+            if (i == 1) {
+                add(Pieces.BLACKPAWN, board);
+                continue;
+            }
+            if (i == 6) {
+                add(Pieces.WHITEPAWN, board);
+                continue;
+            }
+            add(Pieces.NONE, board);
+        }
+        return board;
+    }
```
continue가 많아서 지저분해 보입니다. 다음에는 이 부분을 깔끔하게 개선해 주세요.

honux77 replied:
1, 6 도 매직 넘버 입니다. 제거해 주세요.

honux77 commented on src/main/java/piece/Pawn.java:
```diff
@@ -0,0 +1,24 @@
+        this.logo = pieces.getPresentation();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Pawn pawn = (Pawn) o;
+        return Objects.equals(color, pawn.color);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(color);
+    }
+}
```
불필요한 코드 같네요. 확인해 보고 불필요하면 제거해 주세요.

horanga replied:
현재 코드에서 불필요해서 제거했습니다!

honux77 commented on src/main/java/piece/Pieces.java:
```diff
@@ -0,0 +1,26 @@
+package piece;
+
+public enum Pieces {
+
+    BLACKPAWN("black", "P"),
+
+    WHITEPAWN("white", "p"),
+    NONE("NONE", ".");
```
Piece 라는 enum 타입에 black이라는 다른 정보가 포함되어 있습니다. 클래스의 원래 의미와 맞지 않다고 생각합니다.

horanga replied:
그렇게 한 이유는 

https://techblog.woowahan.com/2527/

여기서 좀 영감을 받아서 이번에 적용해보았습니다. 


![2](https://github.com/codesquad-members-2024/be-chess/assets/148988364/3a55b3e6-5695-4121-88f2-6f9a61e1605f)
![3](https://github.com/codesquad-members-2024/be-chess/assets/148988364/7c95276b-91e4-4541-9d48-0e073883d270)

![사진1](https://github.com/codesquad-members-2024/be-chess/assets/148988364/7eee8fd7-a48e-4a0f-a34e-68ddc74a833a)

여기 사진을 보면 어떤 변수를 놓고 그 변수를 나타내는 데이터들을 연결해주는 용도로 enum을 사용했는데, 추후에 피스들에 정보들이 계속 늘어날 것같아서 이렇게 적용을 해봤는데 저 사진과는 다른 방식으로 쓰여진걸까요? 


이 내용은 step-3에 적용해서 pr에 올렸는데 step-2를 수정하면서 step-3내용까지 같이 커밋을 하면서, step-2 pr에는 이러한 내용을 올리지 못했습니다.

honux77 replied:
저 방식은 같은 데이터의 다른 표현이구요. 지금은 검정과 폰은 직접적인 연관이 없죠. 잘못된 응용이라고 생각해요.

horanga replied:
이 내용도 수정하겠습니다!

honux77 reviewed (changes requested):
전반적으로 요구사항보다 지나치게 많은 내용이 코드에 담겨있어요. 
다음 단계를 고려하지 말고 주어진 단계에서 가장 단순한 코드가 될 수 있도록 재작성해 주세요. 

수정 완료 후에는 코멘트도 하나 추가해 주세요.

horanga committed:
refactor: board 클래스 continue제거
중복 코드 제거를 위해서 if else 구조로 변경

horanga committed:
refactor: Piecs 이넘 클래스에서 color제거
color를 제거하고 직접 인자로 전달하도록 함. 기존 enum 클래스에 너무 많은 책임이 부여돼서 수정함
pawn 생성자에서 enum을 쓰지 않고 String 상수를 전달하기 때문에 예외 처리를 했음

horanga committed:
Revert "refactor: board 클래스 continue제거"

This reverts commit 7726043e08fbc208790ad15b06148b5e712d2999.

horanga committed:
refactor: Piecs 이넘 클래스에서 color제거
color를 제거하고 직접 인자로 전달하도록 함. 기존 enum 클래스에 너무 많은 책임이 부여돼서 수정함
pawn 생성자에서 enum을 쓰지 않고 String 상수를 전달하기 때문에 예외 처리를 했음

horanga committed:
refactor: 매직 넘버 제거
테스트에서도 매직넘버 상수로 변환

horanga committed:
refactor: pawn에 들어오는 색 예외처리
enum을 쓰지 않고 String으로 전달하기로 함
enum을 쓰면 타입 안전성을 얻을 수 있지만 null 체크를 해줘야 해서 큰 이점이 없다고 판단함
String에서 하얀색과 검은색만 받을 수 있도록 예외처리를 했음

horanga commented:
변경사항 최종 수정했습니다!

color를 enum으로 만들려다 enum도 null체크를 해야 한다고 들어서 그렇다면 그냥 String으로 black과 white만 받게 하고 그렇지 않으면 예외처리를 하는 것도 괜찮을 거 같아서 색은 상수로 정의해두었습니다. 데이터가 검은색과 하얀색 둘뿐이라서 상수로 처리해도 괜찮을 거 같다고 생각했습니다.

horanga committed:
refactor: board 클래스 continue제거
중복 코드 제거를 위해서 if else 구조로 변경

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,63 @@
+
+public class Board {
+
+    private static List<List<Piece>> BOARD = intialize();
```
```suggestion
    private static List<List<Piece>> board = intialize();
```

상수가 아닌 객체는 대문자로 하지 않습니다.

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -0,0 +1,63 @@
+    public static final String WHITE = "white";
+
+
+    private static List<List<Piece>> intialize() {
```
```suggestion
    private static void intialize() {
```

멤버변수가 있는데 리턴할 필요가 없겠죠?

horanga replied:
내용 수정했습니다! intialize를 여러번 사용할 필요가 없도록 메서드에서 list를 리턴하고 그걸 멤버변수로 캐싱해두고 get하는 방식을 썼었는데, 메서드 가독성이 좋지 않은거 같아서 멤버변수에 직접 add 해준는 방식을 썼습니다. initialize를 외부에서 계속 쓰게 하는 것도 좋은 선택지가 아닌 거 같아서 static 블록으로 처음에 실행하도록 했습니다.

horanga committed:
refactor: Board 클래스 수정
initalize 메서드가 멤버변수를 이용하도록 리턴타입을 없앰

honux77 commented:
고치느라 수고하셨어요! 👍
